### ISSUE TITLE
model_validate() leads to an uncontrollable increase in memory

### ISSUE THREAD

**Issue by devandreassimmeth** [pydantic/pydantic#8520]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

We are using a combination of FastAPI, Pydantic and Beanie to implement several backends for our Angular frontend. Each of our MongoDB collections use schema-versioning, so we can save multiple document-schemas in one collection and transform it into one unified schema for our frontend.

To do that, we load our data out of the database using the AsyncIOMotorCollection. After that, the dict is mapped from it's current schema-version into the newest and afterwards with "model_validate()" parsed into the Beanie/Pydantic class. Example for that:

```
class MyEntity(Document):
  @classmethod
  async def find_by_id(cls, id: UUID) -> Optional["DocType"]:
    collection = cls.__get_motor_collection()  # Here we get the AsyncIOMotorCollection

    document = await collection.find_one({"_id": id})
    if document is None:
      return None

    return await cls.map_to_newest_schema_version(document)


  @staticmethod
  async def map_to_newest_schema_version(
          document: dict,
      ) -> Optional["MyDataClass"]:
    if document is None:
      return None

    schema_version = document.get("_schema_version", 0)
    if schema_version == 0:
      return MyDataClass.model_validate(
        await MyDataClass.map_from_zero_to_one(document),
      )
    elif schema_version == 1:
      return MyDataClass.model_validate(document)

    raise NotImplementedError()
```

The documents in the DB are pretty big and so are the objects. 

However, a few days ago we got an alarm on our server that the Memory usage of one of our services increased to it's maximum and we found out that the memory-usage of our service constantly increased over weeks/months. 
Because of that we started to search, where this constant increase comes from. After a longer search, we started to use `memory-profiler` in our project to find the code-line where the memory increase is triggered. And while we were doing that, we found out that for reasons we can't explain, some times when we call the pydantic method "model.validate()", our memory increases by mostly the same amount of data and this data is never released again until the service is re-started. So the memory usages increases constantly over time. Here a picture of that increase:

![image](https://github.com/pydantic/pydantic/assets/98212488/03fe01a2-b43e-49f2-85ea-dc3008e58b4d)

Now the question is: Has anyone also faced this behaviour and/or does know a solution how to solve that? Or is it also maybe a bug?

Thanks for help and your answers! 
Andreas :)

### Python, Pydantic & OS Version

```Text
FastAPI: Version 0.108.0
Pydantic: Version 2.5.3
Beanie: Version 1.24.0

Python: 3.10.8

Machine: Windows-PC (Windows 11), but same error also in Docker-Container with "python:3.10-slim-buster" image
```

**Comment by sydney-runkle** [comment]
@devandreassimmeth,

Thanks for reporting this. Are you able to provide a reproducible code snippet that we can use to investigate this bug? Thanks!

**Comment by devandreassimmeth** [comment]
Hello,

I tried now to create an application to reproduce the bug I described. During this development it turned out more and more that it is not the Pydantic function that is the reason for the memory allocation, but that this allocation already happens during the database call using Beanie/Motor. I can also reproduce this with my example application. Because of that we created an issue on the MingoDB Motor issue board (https://jira.mongodb.org/projects/MOTOR/issues/MOTOR-1231?filter=allopenissues) and I think this issue can be closed. Sorry for the inconvenience!
